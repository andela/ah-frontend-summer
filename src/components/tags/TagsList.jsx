import React from 'react';
import PropTypes from 'prop-types';
import { Container, Divider, Grid, Header } from 'semantic-ui-react';
import TagCard from './TagCard';
import '../../styles/tags.scss';

const TagsList = (props) => {
    const {
        tags,
    } = props;
    return(
        <Container>
            <div className="article-list">
                <Header color="teal" as="h1">
                    Tags
                </Header>
                
                <Grid stackable divided="vertically">
                    <Grid.Row columns={2}>
                        <Grid.Column>
                            <Divider />
                            <div className="tags-container">
                                {tags.map(tag => (
                                    <TagCard
                                        key={tag}
                                        tag={tag}
                                    />
                                ))}
                            </div>
                        </Grid.Column>
                    </Grid.Row>
                </Grid>
            </div>
        </Container>
    );
};

TagsList.propTypes = {
    tags: PropTypes.array.isRequired,
};

export default TagsList;
