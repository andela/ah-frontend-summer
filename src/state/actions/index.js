/**
 * This module contains all actions for the project
 * They are further split into synchronous (trigger state update immediately)
 * and asynchronous actions (trigger an operation that will ultimately trigger a synchronous action)
 **/
