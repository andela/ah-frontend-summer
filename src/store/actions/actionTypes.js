export const SIGNUP_USER_STARTED = "SIGNUP_USER_STARTED";
export const SIGNUP_USER_SUCCEEDED = "SIGNUP_USER_SUCCEDED";
export const SIGNUP_USER_FAILED = "SIGNUP_USER_FAILED";

export const VERIFY_USER_ACCOUNT_STARTED = "VERIFY_USER_ACCOUNT_STARTED";
export const VERIFY_USER_ACCOUNT_SUCCEDED = "VERIFY_USER_ACCOUNT_SUCCEDED";
export const VERIFY_USER_ACCOUNT_FAIL = "VERIFY_USER_ACCOUNT_FAIL";
export const DELETE_ARTICLE = "DELETE_ARTICLE";
export const DELETE_ARTICLE_FAIL = "DELETE_ARTICLE_FAIL";

export const GET_ARTICLE_START = "GET_ARTICLE_START";
export const GET_ARTICLE_SUCCESS = "GET_ARTICLE_SUCCESS";
export const GET_ARTICLE_FAIL = "GET_ARTICLE_FAIL";
export const CREATE_ARTICLE_SUCCESS = "CREATE_ARTICLE_SUCCESS";
export const CREATE_ARTICLE_FAIL = "CREATE_ARTICLE_FAIL";
export const UPDATE_ARTICLE_SUCCESS = "UPDATE_ARTICLE_SUCCESS";
export const UPDATE_ARTICLE_FAIL = "UPDATE_ARTICLE_FAIL";

export const ACTION_ENDED = "ACTION_ENDED";

export const FETCH_COMMENTS_ACTION = "FETCH_COMMENTS_ACTION";
export const FETCH_REPLIES_ACTION = "FETCH_REPLIES_ACTION";
export const CREATE_COMMENT_ACTION = "CREATE_COMMENT_ACTION";
export const UPDATE_COMMENT_ACTION = "UPDATE_COMMENT_ACTION";
export const DELETE_COMMENT_ACTION = "DELETE_COMMENT_ACTION";
export const CREATE_REPLY_ACTION = "CREATE_REPLY_ACTION";
export const UPDATE_REPLY_ACTION = "UPDATE_REPLY_ACTION";
export const DELETE_REPLY_ACTION = "DELETE_REPLY_ACTION";
export const GET_PROFILE = "GET_PROFILE";
export const UPDATE_PROFILE = "UPDATE_PROFILE";
export const STARTED_FETCH_USER_PROFILE = "STARTED_FETCH_USER_PROFILE";
export const STARTED_REQUEST_EDIT_PROFILE = "STARTED_REQUEST_EDIT_PROFILE";
export const DISMISS_MESSAGE = "DISMISS_MESSAGE";
export const REQUEST_EDIT_FAILURE = "REQUEST_EDIT_FAILURE";
export const FAILED_FETCH_PROFILE = "FAILED_FETCH_PROFILE";
export const FAILED_UPDATE_PROFILE = "FAILED_UPDATE_PROFILE";
export const STARTED_FETCH_FOLLOWERS = "STARTED_FETCH_FOLLOWERS";
export const STARTED_FETCH_FOLLOWING = "STARTED_FETCH_FOLLOWING";
export const FETCH_TAGS_ACTION = "FETCH_TAGS_ACTION";
export const FETCH_TAG_FILTER_ACTION = "FETCH_TAG_FILTER_ACTION";
