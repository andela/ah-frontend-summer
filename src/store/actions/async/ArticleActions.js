import axios from 'axios';
import { createActionThunk } from 'redux-thunk-actions';


export const FETCH_ARTICLES_ACTION = "FETCH_ARTICLES_ACTION";

const fetchArticles = url => {
    return axios.get(url);
};

export const fetchArticlesAction = createActionThunk(FETCH_ARTICLES_ACTION, fetchArticles, true);
