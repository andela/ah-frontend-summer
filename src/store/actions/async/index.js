import { createActionThunk } from 'redux-thunk-actions';

// placeholder action
export const FETCH_HOME_PAGE_ACTION = "FETCH_HOME_PAGE_ACTION";

export const STARTED = "_STARTED";
export const SUCCEEDED = "_SUCCEEDED";
export const FAILED = "_FAILED";
export const ENDED = "_ENDED";


const fetchHomePage = (url) => {
    // hit our home page endpoint
    return fetch(url);
};

// createActionThunk returns a thunk/ function which you can call with the arguments required
// by your method(which may or may not return a promise) and pass the result to dispatch.
// In this case, something like dispatch(fetchHomePageAction('http://ah-backend-summer-staging.herokuapp.com/'))
// It uses redux thunk underneath and will dispatch the appropriate actions when the call returns
export const fetchHomePageAction = createActionThunk(FETCH_HOME_PAGE_ACTION, fetchHomePage);
