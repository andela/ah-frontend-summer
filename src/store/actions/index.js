/**
 * This module contains all actions for the project
 * They are further split into synchronous (trigger store update immediately)
 * and asynchronous actions (trigger an operation that will ultimately trigger a synchronous action)
 **/

export {
	registerUser,
	verifyUserAccount
} from './auth';
export {
	getArticle,
	deleteArticle,
	createArticle,
	updateArticle
} from './article';
