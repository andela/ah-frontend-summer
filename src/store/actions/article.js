import axios from "axios";

import * as actionTypes from "./actionTypes";

const url = "https://ah-backend-summer-staging.herokuapp.com/api/v1/articles/";

export const articleActionStart = () => {
    return { type: actionTypes.GET_ARTICLE_START };
};

export const getArticleFail = (error) => {
    return { type: actionTypes.GET_ARTICLE_FAIL, error };
};

export const getArticleSuccess = (data) => {
    return {
        type: actionTypes.GET_ARTICLE_SUCCESS,
        data
    };
};

export const deleteArticleAction = (data) => {
    return {
        type: actionTypes.DELETE_ARTICLE,
        data
    };
};

export const deleteArticleFail = (error) => {
    return {
        type: actionTypes.DELETE_ARTICLE_FAIL,
        error
    };
};

export const actionFinised = () => {
    return {
        type: actionTypes.ACTION_ENDED
    };
};

export const deleteArticle = (slug) => {
    return async dispatch => {
        dispatch(articleActionStart());
        let token = localStorage.getItem("token");
        try {
            const response = await axios.delete(`${url}` + slug, {
                headers: {
                    Authorization: `Bearer ${token}`
                }
            });
            dispatch(deleteArticleAction(response.data));
            dispatch(actionFinised());
        } catch (error) {
            dispatch(deleteArticleFail(error.response.data));
        }
    };
};

export const getArticle = (slug) => {
    return async (dispatch) => {
        dispatch(articleActionStart());
        try {
            const response = await axios.get(`${url}${slug}`);
            dispatch(getArticleSuccess(response.data));
        } catch (error) {
            dispatch(getArticleFail(error.response.data));
            dispatch(actionFinised());
        }
    };
};

export const createArticleSuccess = (data) => {
    return {
        type: actionTypes.CREATE_ARTICLE_SUCCESS,
        data
    };
};

export const createArticleFail = (error) => {
    return {
        type: actionTypes.CREATE_ARTICLE_FAIL,
        error
    };
};

export const createArticle = (data) => {
    return async dispatch => {
        let token = localStorage.getItem("token");
        let config = {
            headers: {
                "Authorization": `Bearer ${token}`
            }
        };
        dispatch(articleActionStart())
        try {
            const response = await axios.post(`${url}`, data, config);
            dispatch(createArticleSuccess(response.data));
            dispatch(actionFinised());
        } catch (error) {
            dispatch(createArticleFail(error.response.data));
        }
    }
}

export const updateArticleSuccess = (data) => {
    return {
        type: actionTypes.UPDATE_ARTICLE_SUCCESS,
        data
    };
};

export const updateArticleFail = (error) => {
    return {
        type: actionTypes.UPDATE_ARTICLE_FAIL,
        error
    };
};

export const updateArticle = (slug, data) => {
    return async dispatch => {
        let token = localStorage.getItem("token");
        let config = { headers: { "Authorization": `Bearer ${token}` } };
        dispatch(articleActionStart());
        try {
            const response = await axios.patch(`${url}${slug}`, data, config);
            dispatch(updateArticleSuccess(response.data));
        } catch (error) {
            dispatch(updateArticleFail(error.response.data));
        }
    };
};
