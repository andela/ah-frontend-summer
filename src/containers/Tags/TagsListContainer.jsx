import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import TagsList from '../../components/tags/TagsList';
import fetchTagsAction from '../../store/actions/async/TagsActions';
import TagError from "../../components/tags/TagsError";
import {FAILED, SUCCEEDED} from "../../store/actions/async";

export class TagsListContainer extends Component {

    componentDidMount(){
        this.fetchTags();
    }

    fetchTags = () => {
        const { fetchTagsDispatcher } = this.props;
        fetchTagsDispatcher();
    };

    render() {
        const { status, tags } = this.props;
        this.toRender = null;
        if(status === FAILED){
            this.toRender = <TagError retry={this.fetchTags} />;
        } else if(status === SUCCEEDED){
            this.toRender = (
                <TagsList
                    tags={tags}
                />
            );
        }

        return (
            <div>
                {this.toRender}
            </div>
        );
    }
}

export const mapStateToProps = (state) => {
    return {
        tags: state.tags.tags,
        status: state.tags.status
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        fetchTagsDispatcher : () => dispatch(fetchTagsAction()),
    };
};

TagsListContainer.propTypes = {
    status: PropTypes.string.isRequired,
    fetchTagsDispatcher: PropTypes.func.isRequired,
    tags: PropTypes.array.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(TagsListContainer);
