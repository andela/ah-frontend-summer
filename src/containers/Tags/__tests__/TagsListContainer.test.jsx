import React from 'react';
import { shallow } from 'enzyme';
import { TagsListContainer, mapStateToProps, mapDispatchToProps } from '../TagsListContainer';
import { FAILED, SUCCEEDED } from '../../../store/actions/async';
import TagsList from '../../../components/tags/TagsList';

let fetchTags, handleOnClick, wrapper, status, tags, toRender;

beforeEach(() => {
    fetchTags = jest.fn(),
    handleOnClick = jest.fn(),
    status = SUCCEEDED,
    tags = ["body", "love"];
    toRender = (
        <TagsList
            tags={tags}
            handleOnClick={handleOnClick}
        />
    );
});


describe('TagsListContainer tests', () => {
    let props = {
        status: SUCCEEDED,
        tags: ["body", "love"],
        fetchTagsDispatcher: jest.fn(),
        fetchTagFiltersDispatcher: jest.fn(),
        fetchTagFiltersAction: jest.fn(),
        fetchTagsAction: jest.fn()
    };
    wrapper = shallow(<TagsListContainer {...props} />);

    it('should render TagsListContainer correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should call fetchTagsDispatcher correctly', () => {
        wrapper.instance().fetchTags;
        expect(props.fetchTagsDispatcher).toBeCalled();
    });

    it('should call fetchTags when status is Failed', () => {
        status = FAILED;
        wrapper.setProps({ status });
        expect(props.fetchTagsDispatcher).toBeCalled();
    });

    it('should map state to props', () => {
        let newState = { tags : { tags, status } };
        expect(mapStateToProps( newState )).toEqual({tags, status});
    });

    it('should map dispatch to props', () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch).fetchTagsDispatcher();
        expect(dispatch.mock.calls.length).toBe(1);
    });

});
