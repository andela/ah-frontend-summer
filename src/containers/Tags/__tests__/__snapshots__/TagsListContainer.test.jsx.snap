// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsListContainer tests should render TagsListContainer correctly 1`] = `
<div>
  <TagsList
    tags={
      Array [
        "body",
        "love",
      ]
    }
  />
</div>
`;
